crates/red_knot_test/src/diagnostic.rs

```diff
@@ -0,0 +1,125 @@
+//! Sort and group diagnostics by line number, so they can be correlated with assertions.
+//!
+//! We don't assume that we will get the diagnostics in source order.
+use ruff_source_file::{LineIndex, OneIndexed};
+use ruff_text_size::Ranged;
+use smallvec::SmallVec;
+
+#[derive(Debug)]
+pub(crate) struct SortedDiagnostics<T>(Vec<DiagnosticWithLine<T>>);
+
+impl<T> SortedDiagnostics<T>
+where
+    T: Ranged + Clone,
+{
+    pub(crate) fn new(diagnostics: impl IntoIterator<Item = T>, line_index: &LineIndex) -> Self {
+        let mut diagnostics = diagnostics
+            .into_iter()
+            .map(|diagnostic| DiagnosticWithLine {
+                line: line_index.line_index(diagnostic.range().start()),
+                diagnostic,
+            })
+            .collect::<Vec<_>>();
+        diagnostics.sort_by_key(|diag| diag.line);
+
+        Self(diagnostics)
+    }
+
+    pub(crate) fn iter_lines(&self) -> LineDiagnosticsIterator<T> {
+        LineDiagnosticsIterator {
+            inner: self.0.iter().peekable(),
+        }
+    }
+}
+
+/// Iterator to group sorted diagnostics by line.
+pub(crate) struct LineDiagnosticsIterator<'a, T> {
+    inner: std::iter::Peekable<std::slice::Iter<'a, DiagnosticWithLine<T>>>,
+}
+
+impl<T> Iterator for LineDiagnosticsIterator<'_, T>
+where
+    T: Ranged + Clone,
+{
+    type Item = LineDiagnostics<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let diag = self.inner.next()?;
+        let line = diag.line;
+        let mut diagnostics = DiagnosticVec::new();
+        diagnostics.push(diag.diagnostic.clone());
+        while let Some(diag) = self.inner.peek() {
+            if diag.line == line {
+                diagnostics.push(self.inner.next().unwrap().diagnostic.clone());
+            } else {
+                break;
+            }
+        }
+        Some(LineDiagnostics { line, diagnostics })
+    }
+}
+
+impl<T> std::iter::FusedIterator for LineDiagnosticsIterator<'_, T> where T: Clone + Ranged {}
+
+/// A vector of diagnostics belonging to a single line.
+///
+/// Most lines will have zero or one diagnostics, so we use a [`SmallVec`] optimized for a single
+/// element to avoid most heap vector allocations.
+type DiagnosticVec<T> = SmallVec<[T; 1]>;
```

Comment:
Same comment as for assertions: I think the design you propose is a good one, just not sure if I want to spend the time to do that refactor right now.

Reply:
I did this. I don't think it's really simpler; it ends up being significantly more lines of code, not less, and just as many lifetimes. (Maybe you can observe some ways to simplify it further?)

I think it's worth it for diagnostics regardless because we will often have more than one per line (every un-imported `reveal_type`).